haystack/agents/base.py

```diff
@@ -0,0 +1,271 @@
+            "for a final answer, respond with the `Final Answer:`\n\n"
+            "Use the following format:\n\n"
+            "Question: the question to be answered\n"
+            "Thought: Reason if you have the final answer. If yes, answer the question. If not, find out the missing information needed to answer it.\n"
+            f"Tool: [{tool_names}]\n"
+            "Tool Input: the input for the tool\n"
+            "Observation: the tool will respond with the result\n"
+            "...\n"
+            "Final Answer: the final answer to the question, make it short (1-5 words)\n\n"
+            "Thought, Tool, Tool Input, and Observation steps can be repeated multiple times, but sometimes we can find an answer in the first pass\n"
+            "---\n\n"
+            "Question: $query\n"
+            "Thought: Let's think step-by-step, I first need to "
+        )
+
+    def add_tool(self, tool: Tool):
+        """
+        Add a new tool to the Agent and update the template for the Agent's PromptNode.
+
+        :param tool: The Tool to add to the Agent. Any previously added tool with the same name will be overwritten.
+        """
+        self.tools[tool.name] = tool
+        self.prompt_text = self._generate_prompt_text()
+
+    def run(self, query: str, params: Optional[dict] = None):
+        """
+        Runs the Agent given a query and optional parameters to pass on to the tools used. The result is in the
+        same format as a pipeline's result: a dictionary with a key "answers" containing a List of Answers
+
+        :param query: The search query (for query pipelines only).
+        :param params: A dictionary of parameters that you want to pass to those tools that are pipelines.
+                       To pass a parameter to all nodes in those pipelines, use: `{"top_k": 10}`.
+                       To pass a parameter to targeted nodes in those pipelines, use:
+                        `{"Retriever": {"top_k": 10}, "Reader": {"top_k": 3}}`.
+                        Parameters can only be passed to tools that are pipelines but not nodes.
+        """
+        if len(self.tools) == 0:
+            raise AgentError("Agents without tools cannot be run. Add at least one tool using add_tool().")
+        if params and "max_iterations" in params:
+            remaining_iterations = params["max_iterations"]
+        else:
+            remaining_iterations = self.max_iterations
+
+        transcript = self.prompt_text
+        final_answer = ""
+
+        # In each iteration, choose a tool, generate input for it, and run it until the final answer is found or the
+        # maximum number of iterations is reached.
+        while remaining_iterations > 0:
+            remaining_iterations -= 1
+
+            prompt_template = PromptTemplate("think-step-by-step", transcript, ["query"])
+            pred = self.prompt_node.prompt(prompt_template=prompt_template, query=query)
+            transcript += f"{pred[0]}\n"
```

Comment:
Let's add function `_extract_prediction` - it returns string that we assign to a variable here. So we'll have `pn_pred = self._extract_prediction(pred)` and then we can use pn_pred as a str wherever we need to, including on lines 154 and 156. We thus enable different prediction extraction is subclasses if needed and clean up this code a bit at the same time.